File: engine.py
"""Texas Hold'em hand engine: blinds, betting rounds, side pots, showdown.

`PokerEngine.play_hand` runs one complete hand against a UI object that the
engine calls back into for rendering and for the human's decisions. The engine
owns all the rules; the UI only displays state and collects input.
"""

from __future__ import annotations

import random
from dataclasses import dataclass, field

from bot import Context, decide
from cards import Card, Deck
from evaluator import evaluate
from player import Player


@dataclass
class GameState:
    players: list[Player]          # seated for this hand (all start with chips > 0)
    button: int                    # index into players of the dealer button
    sb: int
    bb: int
    deck: Deck = field(default_factory=Deck)
    community: list[Card] = field(default_factory=list)
    current_bet: int = 0           # highest street_bet this round (amount to match)
    min_raise: int = 0             # minimum legal raise increment right now
    street: str = "preflop"
    log: list[str] = field(default_factory=list)
    hand_no: int = 0

    @property
    def pot(self) -> int:
        return sum(p.total_committed for p in self.players)


@dataclass
class Pot:
    amount: int
    eligible: list[Player]      # in-hand players who may win this layer
    participants: list[Player]  # everyone who contributed to this layer


class PokerEngine:
    def __init__(self, ui, rng: random.Random | None = None) -> None:
        self.ui = ui
        self.rng = rng or random.Random()

    # ---------------------------------------------------------------- helpers
    def _blind_indices(self, n: int, button: int) -> tuple[int, int]:
        if n == 2:
            return button, (button + 1) % n          # SB = button (heads-up)
        return (button + 1) % n, (button + 2) % n     # SB, BB

    def _first_to_act(self, n: int, button: int, preflop: bool) -> int:
        sb, bb = self._blind_indices(n, button)
        if preflop:
            return button if n == 2 else (bb + 1) % n
        return bb if n == 2 else (button + 1) % n      # heads-up: BB acts first postflop

    # ------------------------------------------------------------- one hand
    def play_hand(self, players: list[Player], button: int, sb: int, bb: int,
                  hand_no: int) -> GameState:
        deck = Deck(self.rng)
        state = GameState(players=players, button=button, sb=sb, bb=bb,
                          deck=deck, hand_no=hand_no)
        self.state = state

        for p in players:
            p.reset_for_hand()
            p.last_action = ""

        # Deal two hole cards to each player, starting left of the button.
        n = len(players)
        for _ in range(2):
            for off in range(1, n + 1):
                players[(button + off) % n].hole.extend(deck.deal(1))

        self.ui.hand_start(state)

        self._post_blinds(players, button, sb, bb)

        # Streets: name, community cards to add.
        streets = [("preflop", 0), ("flop", 3), ("turn", 1), ("river", 1)]
        all_in_runout = False

        for name, ncards in streets:
            state.street = name
            if name != "preflop":
                for p in players:
                    p.reset_for_street()
                state.current_bet = 0
                state.min_raise = bb
                state.community.extend(deck.deal(ncards))
                self.ui.on_street(state, name)

            if not all_in_runout:
                start = self._first_to_act(n, button, preflop=(name == "preflop"))
                self._betting_round(players, start)

                if sum(1 for p in players if p.in_hand) == 1:
                    self._award_uncontested(state)
                    return state

                # If at most one player can still act, no more betting: run it out.
                if sum(1 for p in players if p.can_act) <= 1:
                    all_in_runout = True

        # Reached showdown: deal any missing community cards (all-in run-out).
        while len(state.community) < 5:
            state.community.extend(deck.deal(1))

        self._showdown(state)
        return state

    # ------------------------------------------------------------- blinds
    def _post_blinds(self, players: list[Player], button: int, sb: int, bb: int) -> None:
        n = len(players)
        sb_i, bb_i = self._blind_indices(n, button)
        posted_sb = players[sb_i].bet(sb)
        players[sb_i].last_action = f"SB {posted_sb}"
        posted_bb = players[bb_i].bet(bb)
        players[bb_i].last_action = f"BB {posted_bb}"
        self.state.current_bet = bb
        self.state.min_raise = bb
        self.ui.render(self.state)

    # ------------------------------------------------------------- betting
    def _betting_round(self, players: list[Player], start_idx: int) -> None:
        n = len(players)
        idx = start_idx
        guard = 0
        while True:
            steps = 0
            while steps < n and not players[idx].can_act:
                idx = (idx + 1) % n
                steps += 1
            if steps == n:
                return  # nobody can act

            p = players[idx]
            to_call = self.state.current_bet - p.street_bet
            if p.has_acted and to_call == 0:
                return  # action has come back around with everyone matched

            self._act(players, p)
            if sum(1 for q in players if q.in_hand) == 1:
                return

            idx = (idx + 1) % n
            guard += 1
            if guard > 10000:
                return  # safety valve

    def _act(self, players: list[Player], p: Player) -> None:
        st = self.state
        to_call = st.current_bet - p.street_bet
        can_check = to_call == 0
        max_raise_to = p.street_bet + p.chips        # all-in shove total
        min_raise_to = st.current_bet + st.min_raise
        can_raise = max_raise_to > st.current_bet    # has chips beyond a call

        if p.is_human:
            info = {
                "to_call": min(to_call, p.chips),
                "can_check": can_check,
                "can_raise": can_raise,
                "min_raise_to": min(min_raise_to, max_raise_to),
                "max_raise_to": max_raise_to,
                "current_bet": st.current_bet,
            }
            action, amount = self.ui.human_action(st, p, info)
        else:
            opponents = max(1, sum(1 for q in players if q.in_hand and q is not p))
            ctx = Context(
                community=list(st.community),
                current_bet=st.current_bet,
                min_raise=st.min_raise,
                pot=st.pot,
                num_opponents=opponents,
                big_blind=st.bb,
                rng=self.rng,
                street=st.street,
            )
            action, amount = decide(p, ctx)

        self._apply_action(p, action, amount)
        p.has_acted = True
        self.ui.on_action(st, p)

    def _apply_action(self, p: Player, action: str, amount: int) -> None:
        st = self.state
        to_call = st.current_bet - p.street_bet

        if action == "fold":
            # Folding when nothing is owed is pointless; treat as a check.
            if to_call <= 0:
                p.last_action = "CHECK"
                return
            p.folded = True
            p.last_action = "FOLD"
            return

        if action == "check":
            if to_call > 0:                  # can't check facing a bet -> call
                self._do_call(p, to_call)
            else:
                p.last_action = "CHECK"
            return

        if action == "call":
            if to_call <= 0:
                p.last_action = "CHECK"
            else:
                self._do_call(p, to_call)
            return

        if action == "raise":
            target = min(amount, p.street_bet + p.chips)   # cap at all-in
            if target < min(st.current_bet + st.min_raise, p.street_bet + p.chips):
                # Below a full raise and not an all-in shove: bump to legal min.
                if target != p.street_bet + p.chips:
                    target = st.current_bet + st.min_raise
            pay = target - p.street_bet
            p.bet(pay)
            new_bet = p.street_bet
            if new_bet > st.current_bet:
                inc = new_bet - st.current_bet
                if inc >= st.min_raise:
                    st.min_raise = inc
                st.current_bet = new_bet
                p.last_action = ("ALL-IN " if p.all_in else "RAISE ") + str(new_bet)
            else:
                # Could not actually raise (short all-in call).
                p.last_action = ("ALL-IN " if p.all_in else "CALL ") + str(new_bet)
            return

        raise ValueError(f"Unknown action: {action}")

    def _do_call(self, p: Player, to_call: int) -> None:
        paid = p.bet(to_call)
        if p.all_in:
            p.last_action = f"ALL-IN {p.street_bet}"
        else:
            p.last_action = f"CALL {p.street_bet}"

    # ------------------------------------------------------------- pots
    @staticmethod
    def compute_pots(players: list[Player]) -> list[Pot]:
        """Split committed chips into a main pot and any side pots."""
        contribs = {p: p.total_committed for p in players if p.total_committed > 0}
        pots: list[Pot] = []
        while contribs:
            layer = min(contribs.values())
            participants = list(contribs.keys())
            amount = layer * len(participants)
            eligible = [p for p in participants if p.in_hand]
            pots.append(Pot(amount=amount, eligible=eligible, participants=participants))
            new: dict[Player, int] = {}
            for p in participants:
                rest = contribs[p] - layer
                if rest > 0:
                    new[p] = rest
            contribs = new
        return pots

    def _order_from_button(self, contenders: list[Player]) -> list[Player]:
        """Order players by seat starting left of the button (for odd chips)."""
        players = self.state.players
        n = len(players)
        button = self.state.button
        order = [players[(button + off) % n] for off in range(1, n + 1)]
        return [p for p in order if p in contenders]

    def _award_pots(self, state: GameState) -> list[tuple[list[Player], int, int]]:
        awards: list[tuple[list[Player], int, int]] = []
        for pot in self.compute_pots(state.players):
            if not pot.eligible:
                # Uncalled chips: refund to contributors (rare edge case).
                share = pot.amount // len(pot.participants)
                for p in pot.participants:
                    p.chips += share
                continue
            best = max(evaluate(p.hole + state.community).key for p in pot.eligible)
            winners = [p for p in pot.eligible
                       if evaluate(p.hole + state.community).key == best]
            winners = self._order_from_button(winners)
            split = pot.amount // len(winners)
            remainder = pot.amount - split * len(winners)
            for w in winners:
                w.chips += split
            if remainder:
                winners[0].chips += remainder  # odd chip to first seat left of button
            awards.append((winners, split, pot.amount))
        return awards

    # ------------------------------------------------------------- endings
    def _award_uncontested(self, state: GameState) -> None:
        winner = next(p for p in state.players if p.in_hand)
        total = state.pot
        winner.chips += total
        self.ui.on_fold_win(state, winner, total)

    def _showdown(self, state: GameState) -> None:
        in_players = [p for p in state.players if p.in_hand]
        results = [(p, evaluate(p.hole + state.community)) for p in in_players]
        results.sort(key=lambda pr: pr[1].key, reverse=True)
        self.ui.on_showdown(state, results)
        awards = self._award_pots(state)
        self.ui.on_awards(state, awards)

File: src/engine.js
// Hold'em hand engine as a generator state machine.
//
// `playHand` yields events the UI renders (deal, blinds, action, street,
// showdown, awards, foldWin). When the human must act it yields
// { type:"awaitHuman", info } and pauses; the driver resumes with
// gen.next([action, amount]). All rules live here.

import { Deck } from "./cards.js";
import { evaluate } from "./evaluator.js";
import { decide } from "./bot.js";

export class GameState {
  constructor(players, button, sb, bb, deck, handNo = 0) {
    this.players = players;
    this.button = button;
    this.sb = sb;
    this.bb = bb;
    this.deck = deck;
    this.community = [];
    this.currentBet = 0;
    this.minRaise = 0;
    this.street = "preflop";
    this.handNo = handNo;
  }
  get pot() { return this.players.reduce((s, p) => s + p.totalCommitted, 0); }
}

export class PokerEngine {
  constructor(rng = Math.random) { this.rng = rng; }

  _blindIndices(n, button) {
    if (n === 2) return [button, (button + 1) % n];
    return [(button + 1) % n, (button + 2) % n];
  }

  _firstToAct(n, button, preflop) {
    const [, bb] = this._blindIndices(n, button);
    if (preflop) return n === 2 ? button : (bb + 1) % n;
    return n === 2 ? bb : (button + 1) % n;
  }

  *playHand(players, button, sb, bb, handNo) {
    const deck = new Deck(this.rng);
    const state = new GameState(players, button, sb, bb, deck, handNo);
    this.state = state;
    const n = players.length;

    for (const p of players) { p.resetForHand(); p.lastAction = ""; }

    // Deal two hole cards each, starting left of the button.
    for (let k = 0; k < 2; k++)
      for (let off = 1; off <= n; off++)
        players[(button + off) % n].hole.push(...deck.deal(1));
    yield { type: "deal", state };

    this._postBlinds(players, button, sb, bb);
    yield { type: "blinds", state };

    const streets = [["preflop", 0], ["flop", 3], ["turn", 1], ["river", 1]];
    let runout = false;

    for (const [name, ncards] of streets) {
      state.street = name;
      if (name !== "preflop") {
        for (const p of players) p.resetForStreet();
        state.currentBet = 0;
        state.minRaise = bb;
        state.community.push(...deck.deal(ncards));
        yield { type: "street", name, state };
      }

      if (!runout) {
        const start = this._firstToAct(n, button, name === "preflop");
        yield* this._bettingRound(players, start);

        if (players.filter(p => p.inHand).length === 1) {
          const winner = players.find(p => p.inHand);
          const amount = state.pot;
          winner.chips += amount;
          yield { type: "foldWin", winner, amount, state };
          return;
        }
        if (players.filter(p => p.canAct).length <= 1) runout = true;
      }
    }

    while (state.community.length < 5) state.community.push(...deck.deal(1));

    const inPlayers = players.filter(p => p.inHand);
    const results = inPlayers
      .map(p => ({ player: p, hand: evaluate(p.hole.concat(state.community)) }))
      .sort((a, b) => b.hand.score - a.hand.score);
    yield { type: "showdown", results, state };

    const awards = this._awardPots(state);
    yield { type: "awards", awards, state };
  }

  _postBlinds(players, button, sb, bb) {
    const n = players.length;
    const [sbI, bbI] = this._blindIndices(n, button);
    const postedSb = players[sbI].bet(sb);
    players[sbI].lastAction = `SB ${postedSb}`;
    const postedBb = players[bbI].bet(bb);
    players[bbI].lastAction = `BB ${postedBb}`;
    this.state.currentBet = bb;
    this.state.minRaise = bb;
  }

  *_bettingRound(players, startIdx) {
    const n = players.length;
    let idx = startIdx;
    let guard = 0;
    while (true) {
      let steps = 0;
      while (steps < n && !players[idx].canAct) { idx = (idx + 1) % n; steps++; }
      if (steps === n) return;

      const p = players[idx];
      const toCall = this.state.currentBet - p.streetBet;
      if (p.hasActed && toCall === 0) return;

      yield* this._act(players, p);
      if (players.filter(q => q.inHand).length === 1) return;

      idx = (idx + 1) % n;
      if (++guard > 10000) return;
    }
  }

  *_act(players, p) {
    const st = this.state;
    const toCall = st.currentBet - p.streetBet;
    const canCheck = toCall === 0;
    const maxRaiseTo = p.streetBet + p.chips;
    const minRaiseTo = st.currentBet + st.minRaise;
    const canRaise = maxRaiseTo > st.currentBet;

    let action, amount;
    if (p.isHuman) {
      const info = {
        toCall: Math.min(toCall, p.chips),
        canCheck, canRaise,
        minRaiseTo: Math.min(minRaiseTo, maxRaiseTo),
        maxRaiseTo,
        currentBet: st.currentBet,
        pot: st.pot,
      };
      const resp = yield { type: "awaitHuman", player: p, info, state: st };
      [action, amount] = resp;
    } else {
      const opponents = Math.max(1, players.filter(q => q.inHand && q !== p).length);
      const ctx = {
        community: st.community.slice(),
        currentBet: st.currentBet,
        minRaise: st.minRaise,
        pot: st.pot,
        numOpponents: opponents,
        bigBlind: st.bb,
        rng: this.rng,
        street: st.street,
      };
      [action, amount] = decide(p, ctx);
    }

    this._applyAction(p, action, amount);
    p.hasActed = true;
    yield { type: "action", player: p, state: st };
  }

  _applyAction(p, action, amount) {
    const st = this.state;
    const toCall = st.currentBet - p.streetBet;

    if (action === "fold") {
      if (toCall <= 0) { p.lastAction = "CHECK"; return; }
      p.folded = true;
      p.lastAction = "FOLD";
      return;
    }
    if (action === "check") {
      if (toCall > 0) this._doCall(p, toCall);
      else p.lastAction = "CHECK";
      return;
    }
    if (action === "call") {
      if (toCall <= 0) p.lastAction = "CHECK";
      else this._doCall(p, toCall);
      return;
    }
    if (action === "raise") {
      let target = Math.min(amount, p.streetBet + p.chips);
      const fullMin = Math.min(st.currentBet + st.minRaise, p.streetBet + p.chips);
      if (target < fullMin && target !== p.streetBet + p.chips) {
        target = st.currentBet + st.minRaise;
      }
      p.bet(target - p.streetBet);
      const newBet = p.streetBet;
      if (newBet > st.currentBet) {
        const inc = newBet - st.currentBet;
        if (inc >= st.minRaise) st.minRaise = inc;
        st.currentBet = newBet;
        p.lastAction = (p.allIn ? "ALL-IN " : "RAISE ") + newBet;
      } else {
        p.lastAction = (p.allIn ? "ALL-IN " : "CALL ") + newBet;
      }
      return;
    }
    throw new Error("Unknown action: " + action);
  }

  _doCall(p, toCall) {
    p.bet(toCall);
    p.lastAction = (p.allIn ? "ALL-IN " : "CALL ") + p.streetBet;
  }

  // --- pots ---
  static computePots(players) {
    let contribs = new Map();
    for (const p of players) if (p.totalCommitted > 0) contribs.set(p, p.totalCommitted);
    const pots = [];
    while (contribs.size) {
      const layer = Math.min(...contribs.values());
      const participants = [...contribs.keys()];
      const amount = layer * participants.length;
      const eligible = participants.filter(p => p.inHand);
      pots.push({ amount, eligible, participants });
      const next = new Map();
      for (const p of participants) {
        const rest = contribs.get(p) - layer;
        if (rest > 0) next.set(p, rest);
      }
      contribs = next;
    }
    return pots;
  }

  _orderFromButton(contenders) {
    const players = this.state.players;
    const n = players.length;
    const order = [];
    for (let off = 1; off <= n; off++) order.push(players[(this.state.button + off) % n]);
    return order.filter(p => contenders.includes(p));
  }

  _awardPots(state) {
    const awards = [];
    for (const pot of PokerEngine.computePots(state.players)) {
      if (pot.eligible.length === 0) {
        const share = Math.floor(pot.amount / pot.participants.length);
        for (const p of pot.participants) p.chips += share;
        continue;
      }
      let best = -1;
      for (const p of pot.eligible) {
        const s = evaluate(p.hole.concat(state.community)).score;
        if (s > best) best = s;
      }
      let winners = pot.eligible.filter(
        p => evaluate(p.hole.concat(state.community)).score === best);
      winners = this._orderFromButton(winners);
      const split = Math.floor(pot.amount / winners.length);
      const remainder = pot.amount - split * winners.length;
      for (const w of winners) w.chips += split;
      if (remainder) winners[0].chips += remainder;
      awards.push({ winners, split, amount: pot.amount });
    }
    return awards;
  }
}
